Docs/doc 95/doc 107/technical term evaluation parameter

Rachel-Reverie opened the pull request:
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- Adds Evaluation Parameter technical terms page
- Adds Evaluation Parameter technical terms page to the ToC


After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).


### Definition of Done
Please delete options that are not relevant.

- [X] I have made corresponding changes to the documentation


Thank you for submitting!

Rachel-Reverie committed:
- Added initial content to technical terms page for Evaluation Parameters.
- Added technical term page for Evaluation Parameters to the ToC

Rachel-Reverie committed:
- Added tags to technical terms in Evaluation Prameters technical terms page.
- Updated Evaluation Parameter's technical terms page to reflect interaction of Evaluation Parameters, Checkpoints and Actions.

cdkini commented on docs/terms/evaluation_parameter.md:
```diff
@@ -1,15 +1,132 @@
 id: evaluation_parameter
 title: Evaluation Parameter
-hoverText: A parameter used during Validation of an Expectation to reference simple expressions or previousl generated metrics.
+hoverText: A connector to store and retrieve information about parameters used during Validation of an Expectation which reference simple expressions or previously generated metrics.
```
I'm wary of using the word "connector" given it's association with DataConnectors.

Is there a more suitable word we can use here?

joshua-stauffer replied:
I tend to agree with @cdkini -- what about something along the lines of: `A dynamic value used during Validation of an Expectation which is populated by evaluating simple expressions or by referencing previously generated metrics.`

Rachel-Reverie replied:
I'll update this with "A dynamic value used during Validation of an Expectation which is populated by evaluating simple expressions or by referencing previously generated metrics." as suggested.

cdkini commented on docs/terms/evaluation_parameter.md:
```diff
@@ -1,15 +1,132 @@
+
+<CreateHeader/>
+
+When creating expectations through the interactive method, Evaluation Parameters can be provided a temporary initial value.  This will allow you to test Expectations that are meant to rely on values from previous Validation runs before you have actually used them to Validate data.  Say you have created a Pandas dataframe called `my_df` that references a CSV that contains ten rows of data.  (To do ...
```
Should "expectations" be capitalized here?

Rachel-Reverie replied:
Yes; I'll make that correction.

joshua-stauffer commented on docs/terms/evaluation_parameter.md:
```diff
@@ -1,15 +1,132 @@
+
+### Definition
+
+An Evaluation Parameter is a connector to store and retrieve information about parameters used during <TechnicalTag relative="../" tag="validation" text="Validation" /> of an <TechnicalTag relative="../" tag="expectation" text="Expectation" /> which reference simple expressions or previously generated metrics.
```
again, since `connector` happens to be one of technical terms, i think this might be confusing. Off the top of my head, I would recommend `a dynamic value`, but i'm sure there are other options as well.

joshua-stauffer commented on docs/terms/evaluation_parameter.md:
```diff
@@ -1,15 +1,132 @@
+
+### Relationship to other objects
+
+Evaluation Parameters are used in Expectations when Validating data.  <TechnicalTag relative="../" tag="checkpoint" text="Checkpoints" /> use <TechnicalTag relative="../" tag="validation_action" text="Action" /> to store Evaluation Parameters in the <TechnicalTag relative="../" tag="evaluation_parameter_store" text="Evaluation Parameter Store" />.
```
should this be `Checkpoints use an Action`, or `Checkpoints use Actions`?

Rachel-Reverie replied:
Checkpoints use Actions.  I'll update it, thanks for pointing this out!

joshua-stauffer commented on docs/terms/evaluation_parameter.md:
````diff
@@ -1,15 +1,132 @@
+
+For an example of this, see below:
+
+```python title="Python code"
````
since we reference loading the CSV into memory, it might be nice to include that line here as well: 
```
csv_path = 'path/to/my_data.csv'
my_df = great_expectations.read_csv(csv_path)
```

joshua-stauffer commented on docs/terms/evaluation_parameter.md:
````diff
@@ -1,15 +1,132 @@
+    expectation_suite=my_dag_step_config, 
+    evaluation_parameters={"upstream_row_count": upstream_row_count}
+)
+```
````
This approach avoids the use of URNs altogether -- is that intentional? That logic is currently buggy, but a patch is incoming. Otherwise, the previous doc on using eval params is essentially correct. An Evaluation Parameter can be defined by identifying a result in another suite:
```
suite_name = 'my_suite_1'
expectation_name = 'expect_table_row_count_to_equal'
metric = 'observed_value'
eval_param_urn = f'urn:great_expectations:validations:{suite_name}:{expectation_name}.result.{metric}'
validator.expect_table_row_count_to_equal(
   value={
      '$PARAMETER': eval_param_urn
   }
)
```

joshua-stauffer replied:
I think I would recommend a second section of the Use Cases section to address this specific use case

Rachel-Reverie replied:
Added, along with `import great_expectations` for completeness.

joshua-stauffer commented on docs/terms/evaluation_parameter.md:
```diff
@@ -1,15 +1,132 @@
+
+<CreateHeader/>
+
+When creating expectations through the interactive method, Evaluation Parameters can be provided a temporary initial value.  This will allow you to test Expectations that are meant to rely on values from previous Validation runs before you have actually used them to Validate data.  Say you have created a Pandas dataframe called `my_df` that references a CSV that contains ten rows of data.  (To do ...
```
Since this is the `use cases` section, I would clarify that this use case is `Validate an Expectation against a previous run of the same Suite`, or something along those lines. In contrast, the URN approach allows for validating an expectation against a result in a different suite (useful in part since this means it can be generated from a different data source!)

Rachel-Reverie replied:
I've added reference and an example of how to use an URN to pull values from an Evaluation Parameter Store.

joshua-stauffer commented on docs/terms/evaluation_parameter.md:
```diff
@@ -1,15 +1,132 @@
+
+<ValidateHeader/>
+
+Evaluation Parameters that are configured as part of a Checkpoint's Expectations will be used without further interaction from you.  Additionally, Evaluation Parameters will be stored by having the `StoreEvaluationParametersAction` subclass of the `ValidationAction` class defined in a Checkpoint configuration's `action_list`.
```
Checkpoints also have an object named evaluation_parameters -- could we add some detail about that? I haven't seen anyone use them that way, and would be curious to know more.

Rachel-Reverie replied:
I'm actually pretty unfamiliar with it myself -- this may be something I'll need to consult with @jcampbell on.

jcampbell replied:
Definitely worth mentioning. The evaluation parameters field in a checkpoint definition makes it possible to provide evaluation parameters at when the checkpoint runs, either via a dictionary or, for example, as an environment variable:

(UNTESTED OFF TOP OF HEAD EXAMPLE)

Suppose we have an expectation built like:
```
suite.add_expectation(
  ExpectationConfiguration(
    expectation_type="expect_table_row_count_to_equal",
    kwargs={
        "value": {"$PARAMETER": "upstream_row_count"}
    }
  )
)
```

```
name: my_checkpoint
...
evaluation_parameters:
  upstream_row_count: $MY_ENV_VAR
```

```
context.run_checkpoint('foo', evaluation_parameters={"upstream_row_count": 42})
```

Rachel-Reverie replied:
I've verified that these examples work and have added them into the document.

joshua-stauffer commented on docs/terms/evaluation_parameter.md:
```diff
@@ -1,15 +1,132 @@
+
+### How to create
+
+An Evaluation Parameter is "created" when a parameter is referenced in the results of another Expectation Suite's Validation.  To store them, define a `StoreEvaluationParametersAction` subclass of the `ValidationAction` class in a Checkpoint configuration's `action_list`, and run that Checkpoint.
```
I think this could be more precise - an evaluation param is created when an expectation references an eval param, right? I'm still working on my mental model for this! Also, this action is created by default with `ge init`, so might be worth mentioning that it typically doesn't need to be added.

Rachel-Reverie replied:
Strictly speaking, I think it's created when it references a Metric generated by the results of a previous Validation of the current (or another) Expectation Suite, or when an expression defined as an Evaluation Parameter is evaluated prior to Validation of the current Expectation Suite.  I'll update this section to make that clearer..

joshua-stauffer commented on docs/terms/evaluation_parameter.md:
```diff
@@ -1,15 +1,132 @@
+
+An Evaluation Parameter is "created" when a parameter is referenced in the results of another Expectation Suite's Validation.  To store them, define a `StoreEvaluationParametersAction` subclass of the `ValidationAction` class in a Checkpoint configuration's `action_list`, and run that Checkpoint.
+
+Evaluation Parameters are *referenced* by making a dictionary with the `$PARAMETER` key.  The value for this key will be the Evaluation Parameter's expression, which will be evaluated at run time and replaced with the value described by said expression.
```
I think we should include URNs here as well

Rachel-Reverie replied:
I've added a reference to them here as well (and done a bit of an overhaul of this section for clarity's sake).

joshua-stauffer commented:
This is fantastic, @Rachel-Reverie! I love how you capture this complex process in straightforward and clear prose. I left some thoughts and questions to clarify a few corners, particularly around the use of URNs in eval params. thanks!

Rachel-Reverie committed:
- Content updates for clarity and completeness per resolved discussions on GitHub.

jcampbell commented on docs/term_tags/terms.json:
```diff
@@ -61,7 +61,7 @@
   "evaluation_parameter_store": {
     "term": "Evaluation Parameter Store",
-    "definition": "A connector to store and retrieve information about parameters used during Validation of an Expectation which reference simple expressions or previously generated metrics.",
+    "definition": "A dynamic value used during Validation of an Expectation which is populated by evaluating simple expressions or by referencing previously generated metrics.",
```
I think this is the definition for "Evaluation Parameter" vice "Evaluation Parameter Store"

Rachel-Reverie replied:
It is.  I've rolled the definition of Evaluation Parameter Store back and changed this to be assigned as the definition of Evaluation Parameter.

jcampbell commented on docs/terms/evaluation_parameter.md:
```diff
@@ -1,15 +1,152 @@
+
+The core of this is a `$PARAMETER : URN` pair. When Great Expectations encounters a `$PARAMETER` flag during validation, it will replace the `URN` with a value retrieved from an Evaluation Parameter Store or Metrics Store.
+
+If you do not have a previous Expectation Suite's Validation Results to reference, however, you can instead provide Evaluation Parameters with a temporary initial value. For example, the interactive method of creating Expectations is based on Validating Expectations against a previous run of the same Expectation Suite.  Since a previous run has not been performed when Expectations are being creat ...
```
We may also want to mention that you can create expectations with evaluation parameters as placeholders by turning off interactive evaluation or adding the expectation configuration directly to the suite.

Rachel-Reverie replied:
I've added a mention of this after the list of methods that can be used while interactive evaluation is enabled.

jcampbell commented on docs/terms/evaluation_parameter.md:
````diff
@@ -1,15 +1,152 @@
+my_df.set_evaluation_parameter("upstream_row_count", 10)
+```
+
+If the Evaluation Parameter's value is set in this way, you do not need to set it again (or define it alongside the use of the `$PARAMETER` key) for future Expectations.
````
We need to update these examples to not reference "my_df" and instead reference a validator-based flow. (Those examples come from the v2 api)

Rachel-Reverie replied:
I've updated them to use a validator, and to explain that the validator can be gotten as part of the interactive process for creating Expectations.  I also tested all of the updated examples (one was broken in both V2 and V3!)

Rachel-Reverie committed:
- Corrected issue where definition of Evaluation Parameter had been assigned as the definition of Evaluation Parameter Store.

Rachel-Reverie committed:
- Added mention of writing Evaluation Parameters based on domain knowledge while interactive evaluation is disabled.

Rachel-Reverie committed:
- Added example for configuring Evaluation Parameter to pull value from environment varibale.
- Added example for passing Evaluation Parameter to Checkpoint as a dictionary.

Rachel-Reverie committed:
- Updated all "my_df." examples from V2 to "validator." examples compliant with V3
- Added the examples regarding the `evaluation_parameters` parameter in Checkpoints (but they still need tested).

Rachel-Reverie committed:
- corrected missing closed parenthesis in one of the code examples.

Rachel-Reverie committed:
Merge branch 'universal-map-docs-update' into DOCS/DOC-95/DOC-107/technical_term_evaluation_parameter

Rachel-Reverie committed:
Merge branch 'universal-map-docs-update' into DOCS/DOC-95/DOC-107/technical_term_evaluation_parameter

Rachel-Reverie committed:
-Updated technical term tag for Action after merging in updates.

joshua-stauffer reviewed (approved):
This looks great -- I really appreciate the extra details around using eval params through the yaml config. thanks!

Rachel-Reverie committed:
Merge branch 'universal-map-docs-update' into DOCS/DOC-95/DOC-107/technical_term_evaluation_parameter
